engine/src/flutter/shell/platform/darwin/ios/framework/Source/UITextInput+TextRange.swift

```diff
@@ -0,0 +1,221 @@
+  /// UTF8 encoding. See: https://github.com/flutter/flutter/issues/111494.
+  @objc func safeRange(in document: UITextInput) -> TextRange {
+    let fullText = document.fullText
+    let start = self.start.index
+    let end = self.end.index
+    let startsWithTrailingSurrogate =
+      start > 0 && Unicode.UTF16.isTrailSurrogate(fullText.utf16[start])
+    let endsWithLeadingSurrogate =
+      isEmpty
+      ? startsWithTrailingSurrogate
+      : Unicode.UTF16.isLeadSurrogate(fullText.utf16[end - 1])
+
+    let newStart = startsWithTrailingSurrogate ? start - 1 : start
+    let newEnd = endsWithLeadingSurrogate ? end + 1 : end
+    return TextRange(
+      NSRange: NSRange(location: Int(newStart), length: newStart.distance(to: newEnd)))
+  }
+
+  override var description: String {
+    "TextRange(\(range))"
+  }
+}
+
+// MARK: Class methods for initializing in Objective-C
+
+@objc extension TextPosition {
+  class func position(withIndex index: UInt, affinity: UITextStorageDirection)
+    -> TextPosition
+  {
+    TextPosition(index: index, affinity: affinity)
+  }
+
+  class func position(withIndex index: UInt) -> TextPosition {
+    TextPosition(index: index)
+  }
+}
+
+@objc extension TextRange {
+  class func range(withNSRange range: NSRange) -> TextRange {
+    TextRange(NSRange: range)
+  }
+}
+
+// MARK: Convenience extensions on UITextRange and UITextPosition
+
+@objc extension UITextRange {
+  /// The UTF16 range this FlutterTextRange represents.
+  ///
+  /// This computed property throws if the receiver is not a FlutterTextRange.
+  /// This is a left closed right open interval.
+  var range: NSRange { (self as! TextRange).nsRange }
+}
+
+@objc extension UITextPosition {
+  /// The offset from the start of the document to this FlutterTextPosition, in UTF16 code units.
+  ///
+  /// This computed property throws if the receiver is not a FlutterTextRange.
+  var index: UInt { (self as! TextPosition).offset }
+
+  func compare(to other: UITextPosition) -> ComparisonResult {
+    let otherIndex = other.index
+    let index = self.index
+    if otherIndex == index {
+      return .orderedSame
+    } else if otherIndex > index {
+      return .orderedAscending
+    }
+    return .orderedDescending
+  }
+
+  func offset(by offset: Int, inDocument document: UITextInput)
+    -> TextPosition?
+  {
+    let newOffset = offset + Int(index)
+    if newOffset < 0 || newOffset > (document.fullText as NSString).length {
+      return nil
+    }
+    return TextPosition(
+      index: UInt(newOffset), affinity: document.selectionAffinity ?? .forward)
+  }
+}
+
+// MARK: Convenience extensions
+
+extension String.UTF16View {
+  subscript(index: UInt) -> Unicode.UTF16.CodeUnit {
+    self[self.index(startIndex, offsetBy: Int(index))]
+  }
+}
+
+@objc extension NSString {
+  // Get the UTF16 range for backspace deletion, when the caret is placed at
```

Comment:
is this new logic? why do we need them?

Reply:
Yes this is new. I believe this matches the backspace behavior of TextKit2.